web/src/pages/post/[id].tsx

```diff
@@ -0,0 +1,246 @@
+import { AnimatePresence, motion } from "framer-motion";
+import { useRouter } from "next/router";
+import { useEffect, useMemo, useState } from "react";
+import { toast } from "react-hot-toast";
+import { MdArrowForward, MdLibraryAdd } from "react-icons/md";
+
+import CircularProgress from "../../components/Feedback/CircularProgress";
+import Button from "../../components/Inputs/Button";
+import ImageList from "../../components/Layout/ImageList";
+import AddToCollectionPanel from "../../components/Panels/AddToCollectionPanel";
+import CustomImage from "../../components/Surfaces/CustomImage";
+import Author from "../../components/Widgets/Author";
+import FloatingImageActions from "../../components/Widgets/FloatingImageActions";
+import LikesCounter from "../../components/Widgets/LikesCounter";
+import { useMediaQuery } from "../../hooks/useMediaQuery";
+import { getPost, searchPosts } from "../../requests/fetch";
+import { likePost, unlikePost } from "../../requests/mutate";
+import {
+  transitions,
+  transitionVariants,
+} from "../../styles/motion-definitions";
+import { getAuthSession } from "../../utils/storage";
+
+export default function Post() {
+  //#region Hooks
+
+  const router = useRouter();
+
+  const md = useMediaQuery("(min-width: 768px)");
+
+  const [isAddToCollectionPanelOpen, setIsAddToCollectionPanelOpen] =
+    useState(false);
+  // Tracks a local version of whether the post is liked or not which is updated immediately
+  const [localIsLiked, setLocalIsLiked] = useState(false);
+
+  const { data: postData, isLoading: isPostLoading } = useQuery({
+    queryKey: ["post", router.query.id],
+    queryFn: () => getPost(router.query.id as string),
+    onError: (err: Error) => {
+      toast.error(err.message);
+    },
+    enabled: !!router.query.id,
+  });
+
+  const { data: featuredPostsData, isLoading: areFeaturedPostsLoading } =
+    useQuery({
+      queryKey: ["featured_posts"],
+      queryFn: () => searchPosts({ search: "", limit: 9 }),
+      onError: (err: Error) => {
+        toast.error("Failed to load featured posts");
+      },
+    });
+
+  const { mutate: mutateLike, isLoading: isLiking } = useMutation({
+    mutationKey: ["like"],
+    mutationFn: () => likePost(postData?.id ?? ""),
+    onError: (err: Error) => {
+      setLocalIsLiked(false); // Revert the local state on failure
+      toast.error(err.message);
+    },
+  });
+
+  const { mutate: mutateUnlike, isLoading: isUnliking } = useMutation({
+    mutationKey: ["unlike"],
+    mutationFn: () => unlikePost(postData?.id ?? ""),
+    onError: (err: Error) => {
+      setLocalIsLiked(true); // Revert the local state on failure
+      toast.error(err.message);
+    },
+  });
+
+  const isLiked = useMemo(() => {
+    return !!postData?.likes.find(
+      (like) => like.userId === getAuthSession().userId
+    );
+  }, [postData]);
+
+  useEffect(() => {
+    setLocalIsLiked(isLiked);
+  }, [isLiked]);
```

Comment:
I would just move the `setLocalIsLiked(isLiked)` into the `useMemo()` above so you don't need this effect.